File: src-julia/Problems/DirichletWaveguide2D.jl
module DirichletWaveguide2DM

include("FreeParticle.jl")
using .FreeParticleM: FreeParticle
include("DirichletWell1D.jl")
using .DirichletWell1DM: DirichletWell1D

type DirichletWaveguide2D
    aY :: Float64
    bY :: Float64
    maxn :: Int

    freeX :: FreeParticle
    wellY :: DirichletWell1D

    # wavevectors :: Array{Complex{Float64}}
    # eigenenergies :: Array{Float64}
    # eigenstates :: Array{Function}
    # deigenstates :: Array{Function}

    # greensFunctionHelmholtz :: Function
    greensFunctionHelmholtzDys :: Function


    function DirichletWaveguide2D (aY :: Float64, bY :: Float64, maxn :: Int)
        this = new ()
        this.aY = aY
        this.bY = bY
        this.maxn = maxn

        this.freeX = FreeParticle()
        this.wellY = DirichletWell1D(aY, bY, maxn)

        this.greensFunctionHelmholtzDys = function (energy :: Complex{Float64})
            return function (x :: Float64, y :: Float64, xs :: Float64, ys :: Float64)
                res = complex(0.0)
                for m = 1: maxn
                    gf = this.freeX.greensFunctionHelmholtz(complex(energy - this.wellY.eigenenergies[m]))
                    res += this.wellY.eigenstates[m](y) * this.wellY.deigenstates[m](ys) * gf(x, xs)
                end
                return res
            end
        end

        return this
    end

end


end

File: src/scattering/problems/dirichlet_waveguide_2d.py
from sys import stderr

import scipy.constants as sc

import numpy as np
from numpy import sqrt, exp, cos
from numpy import complex256 as complex
from numpy import float128 as real
from numpy.testing import assert_almost_equal
from scattering.problems.dirichlet_well_1d import DirichletWell1D
from scattering.problems.free_particle import FreeParticle

from scattering.problems.neumann_well_1d import NeumannWell1D
from scattering.tools import integrate_complex
from scattering.tools.tools import inner_product_L2

class DirichletWaveguide2D:
    def __init__(self, aY: real, bY: real, maxn: int):
        self.aY = aY
        self.bY = bY
        self.maxn = maxn

        self.wellY = DirichletWell1D(aY, bY, maxn)
        self.freeX = FreeParticle()

        self.eigenenergies = None # ???
        self.eigenfunctions = None # ???

    # TODO test
    def greens_function_helmholtz_dy(self, energy):
        def fun(x, y, xs, ys, maxn=self.maxn):
            res = complex(0.0)
            for m in range(1, maxn):
                gf = self.freeX.greens_function_helmholtz(energy - self.wellY.eigenenergies[m])
                # NOTE: no conjugation, wavefunctions are real
                res += self.wellY.eigenfunctions[m](y) * self.wellY.deigenfunctions[m](ys) * gf(x, xs)
            return res
        return fun

    def test(self):
        raise NotImplementedError

if __name__ == '__main__':
    dw = DirichletWaveguide2D(-1, 0, 1000)
    gf_dy = dw.greens_function_helmholtz_dy(100.0)
    print(gf_dy(0.0, -1.0, 0.0, 0.0))
    print(gf_dy(1.0, -1.0, 0.0, 0.0))
    print(gf_dy(-1.0, -1.0, 0.0, 0.0))

    print(gf_dy(0.0, -0.5, 0.0, 0.0))
    print(gf_dy(1.0, -0.5, 0.0, 0.0))
    print(gf_dy(-1.0, -0.5, 0.0, 0.0))

    print(gf_dy(0.0, -0.001, 0.0, 0.0))
    print(gf_dy(1.0, 0.0, 0.0, 0.0))
    print(gf_dy(-1.0, 0.0, 0.0, 0.0))